File: Homework 2/Cheatsheet/Cheatsheet.cs
using System;
using System.Collections;
using System.Collections.Generic;

class Program
{
    static void Main()
    {
        // Array
        int[] array = new int[] { 1, 2, 3, 4, 5 };
        // Loop
        foreach (int num in array)
        {
            Console.WriteLine(num);
            // Break/Continue can be used here
        }
        // Add
        int[] newArray = new int[array.Length + 1];
        Array.Copy(array, newArray, array.Length);
        newArray[newArray.Length - 1] = 6;
        // Remove
        newArray = Array.FindAll(newArray, val => val != 3);
        // Get
        int element = newArray[2];
        // Set
        newArray[1] = 10;
        // Check existence of key/value
        bool exists = Array.Exists(newArray, val => val == 5);
//###########################################################
        // List
        List<int> list = new List<int> { 1, 2, 3, 4, 5 };
        // Loop
        foreach (int num in list)
        {
            Console.WriteLine(num);
            // Break/Continue can be used here
        }
        // Add
        list.Add(6);
        // Remove
        list.Remove(3);
        // Get
        int listElement = list[2];
        // Set
        list[1] = 10;
        // Check existence of key/value
        bool listExists = list.Contains(5);
//###########################################################

        // Dictionary
        Dictionary<string, int> dictionary = new Dictionary<string, int>
        {
            { "one", 1 },
            { "two", 2 },
            { "three", 3 }
        };
        // Loop
        foreach (var item in dictionary)
        {
            Console.WriteLine($"{item.Key}: {item.Value}");
            // Break/Continue can be used here
        }
        // Add
        dictionary.Add("four", 4);
        // Remove
        dictionary.Remove("two");
        // Get
        int dictValue;
        bool keyExists = dictionary.TryGetValue("three", out dictValue);
        // Set
        dictionary["one"] = 10;
        // Check existence of key/value
        bool dictExists = dictionary.ContainsKey("four");
//###########################################################

        // SortedList
        SortedList<int, string> sortedList = new SortedList<int, string>
        {
            { 3, "three" },
            { 1, "one" },
            { 2, "two" }
        };
        // Loop
        foreach (var item in sortedList)
        {
            Console.WriteLine($"{item.Key}: {item.Value}");
            // Break/Continue can be used here
        }
        // Add
        sortedList.Add(4, "four");
        // Remove
        sortedList.Remove(2);
        // Get
        string sortedValue = sortedList[3];
        // Set
        sortedList[1] = "updated one";
        // Check existence of key/value
        bool sortedExists = sortedList.ContainsValue("four");
//###########################################################

        // HashSet
        HashSet<int> hashSet = new HashSet<int> { 1, 2, 3, 4, 5 };
        // Loop
        foreach (int num in hashSet)
        {
            Console.WriteLine(num);
            // Break/Continue can be used here
        }
        // Add
        hashSet.Add(6);
        // Remove
        hashSet.Remove(3);
        // Check existence of key/value
        bool hashSetExists = hashSet.Contains(5);

        // SortedSet
        SortedSet<int> sortedSet = new SortedSet<int> { 3, 1, 2, 5, 4 };
        // Loop
        foreach (int num in sortedSet)
        {
            Console.WriteLine(num);
            // Break/Continue can be used here
        }
        // Add
        sortedSet.Add(6);
        // Remove
        sortedSet.Remove(3);
        // Check existence of key/value
        bool sortedSetExists = sortedSet.Contains(5);
//###########################################################

        // Queue
        Queue<string> queue = new Queue<string>();
        queue.Enqueue("first");
        queue.Enqueue("second");
        queue.Enqueue("third");
        // Loop
        foreach (string item in queue)
        {
            Console.WriteLine(item);
            // Break/Continue can be used here
        }
        // Add
        queue.Enqueue("fourth");
        // Remove
        string dequeuedItem = queue.Dequeue();
        // Check existence of key/value
        queue.Contains(5);

//###########################################################

        // Stack
        Stack<string> stack = new Stack<string>();
        stack.Push("first");
        stack.Push("second");
        stack.Push("third");
        // Loop
        foreach (string item in stack)
        {
            Console.WriteLine(item);
            // Break/Continue can be used here
        }
        // Add
        stack.Push("fourth");
        // Remove
        string poppedItem = stack.Pop();
        // Check existence of key/value
        stack.Contains(1);
//###########################################################

        // LinkedList
        LinkedList<int> linkedList = new LinkedList<int>();
        linkedList.AddLast(1);
        linkedList.AddLast(2);
        linkedList.AddLast(3);
        // Loop
        foreach (int num in linkedList)
        {
            Console.WriteLine(num);
            // Break/Continue can be used here
        }
        // Add
        linkedList.AddLast(4);
        // Remove
        linkedList.Remove(2);
        // Check existence of key/value 
        linkedList.Contains(9);
        
    }
}

File: Homework 2/Cheatsheet/Cheatsheet.js

//Array
const array = [1, 2, 3, 4, 5];
// Loop (for loop with break/continue)
for (let i = 0; i < array.length; i++) {
    if (array[i] === 3) {
        break; // Exit the loop
    }
    if (array[i] === 2) {
        continue; // Skip the rest of the loop and continue to the next iteration
    }
    console.log(array[i]);
}

// Add/Remove/Get/Set/Check Existence
array.push(6); // Add an element to the end
array.pop();  // Remove an element from the end
array[0];     // Get an element by index
array[1] = 10; // Set the value of an element
array.includes(3); // Check if an element exists in the array (returns true/false)

//#####################################################################
// Dictionary
const dictionary = {
    key1: "value1",
    key2: "value2"
};

// Loop through keys/values
for (const key in dictionary) {
    if (dictionary.hasOwnProperty(key)) {
        const value = dictionary[key];
        console.log(key, value);
    }
}

// Add/Remove/Get/Set/Check Existence
dictionary.key3 = "value3"; // Add a key-value pair
delete dictionary.key2;     // Remove a key-value pair
dictionary.key1;             // Get the value by key
dictionary.key1 = "new value"; // Set the value of a key
dictionary.hasOwnProperty("key1"); // Check if a key exists (returns true/false)
//#####################################################################

//SORTED LIST 
const sortedList = [1, 2, 4, 4, 3];
sortedList.sort();

// LOOP
sortedList.forEach(i => console.log(i));

// ADD
sortedList.push(6);
sortedList.sort();

// REMOVE
index = sortedList.indexOf(2);
index > -1 && sortedList.splice(index, 1);

// GET
array[1];

// SET
index = sortedList.indexOf(3);
index > -1 && sortedList.splice(index, 1, 5);
sortedList.sort();

// EXISTS KEY/VALUE
sortedList.includes(2);


//##########################################################
const hashSet = new Set([1, 2, 3]);

// Loop (for...of loop)
for (const item of hashSet) {
    console.log(item);
}

// Add/Remove/Check Existence
hashSet.add(4);    // Add an element
hashSet.delete(2); // Remove an element
hashSet.has(3);    // Check if an element exists (returns true/false)

//##########################################################
// Queue
const queue = [];
//Loop
while (queue.length > 0) {
    const element = queue.shift();
    // Do something with element
}

// Enqueue/Dequeue
queue.push(1);        // Enqueue an element
const dequeued = queue.shift(); // Dequeue an element from the front
const element = queue[0];//Get (Peek, without removing):



//##########################################################
//stack
const stack = [];

//Loop
while (stack.length > 0) {
    const element = stack.pop();
    
}

// Push/Pop
stack.push(1);      // Push an element onto the stack
const popped = stack.pop(); // Pop an element from the top of the stack
const element = stack[stack.length - 1]; // Get (Peek, without removing):

stack.includes(3);// EXISTS KEY/VALUE




//##########################################################
//Linkedlist
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
    }

    // Add a node to the end of the linked list
    add(value) {
        const newNode = new Node(value);
        if (!this.head) {
            this.head = newNode;
        } else {
            let current = this.head;
            while (current.next) {
                current = current.next;
            }
            current.next = newNode;
        }
    }

    // Remove a node by its value
    remove(value) {
        if (!this.head) {
            return;
        }
        if (this.head.value === value) {
            this.head = this.head.next;
            return;
        }
        let current = this.head;
        while (current.next) {
            if (current.next.value === value) {
                current.next = current.next.next;
                return;
            }
            current = current.next;
        }
    }

    // Get a node by its index
    get(index) {
        if (index < 0 || !this.head) {
            return null;
        }
        let current = this.head;
        for (let i = 0; i < index; i++) {
            if (!current.next) {
                return null;
            }
            current = current.next;
        }
        return current.value;
    }

    // Set the value of a node at a specific index
    set(index, value) {
        if (index < 0 || !this.head) {
            return false;
        }
        let current = this.head;
        for (let i = 0; i < index; i++) {
            if (!current.next) {
                return false;
            }
            current = current.next;
        }
        current.value = value;
        return true;
    }

    // Loop through the linked list
    loop() {
        let current = this.head;
        while (current) {
            console.log(current.value);
            current = current.next;
        }
    }
}

// Example usage:
const linkedList = new LinkedList();
linkedList.add(1);
linkedList.add(2);
linkedList.add(3);

linkedList.loop(); // Output: 1, 2, 3
linkedList.remove(2);
linkedList.loop(); // Output: 1, 3
console.log(linkedList.get(1)); // Output: 3
linkedList.set(1, 4);
linkedList.loop(); // Output: 1, 4


